modules/passkey/test/utils/safe.ts

```diff
@@ -0,0 +1,75 @@
+import { Hex } from '@noble/curves/abstract/utils'
+import { ethers } from 'hardhat'
+import { Address } from 'hardhat-deploy/types'
+
+const SAFE_TX_TYPEHASH = ethers.keccak256(
+  ethers.toUtf8Bytes(
+    'SafeTx(address to,uint256 value,bytes data,uint8 operation,uint256 safeTxGas,uint256 baseGas,uint256 gasPrice,address gasToken,address refundReceiver,uint256 nonce)',
+  ),
+)
+
+export const DOMAIN_SEPARATOR_TYPEHASH = ethers.keccak256(ethers.toUtf8Bytes('EIP712Domain(uint256 chainId,address verifyingContract)'))
+
+export const buildSafeTransactionData = (safeTx: SafeTransaction, domainSeparator: Hex): Hex => {
+  const safeTxHash = ethers.keccak256(
+    ethers.AbiCoder.defaultAbiCoder().encode(
+      ['bytes32', 'address', 'uint256', 'bytes32', 'uint8', 'uint256', 'uint256', 'uint256', 'address', 'address', 'uint256'],
+      [
+        SAFE_TX_TYPEHASH,
+        safeTx.to,
+        safeTx.value,
+        ethers.keccak256(safeTx.data),
+        safeTx.operation,
+        safeTx.safeTxGas,
+        safeTx.baseGas,
+        safeTx.gasPrice,
+        safeTx.gasToken,
+        safeTx.refundReceiver,
+        safeTx.nonce,
+      ],
+    ),
+  )
+  return ethers.solidityPacked(['bytes1', 'bytes1', 'bytes32', 'bytes32'], ['0x19', '0x01', domainSeparator, safeTxHash])
+}
```

Comment:
nit: this should be implemented with the `ethers.TypedDataEncoder` type, which abstracts a lot of this from you already:

Reply:
```suggestion
export type SafeDomain = {
  verifyingContract: AddressLike
  chainId: BigNumberish
}

export const buildSafeTransactionData = (domain: SafeDomain, safeTx: SafeTransaction): BytesLike => {
  return ethers.TypedDataEncoder.encode(
    domain,
    {
      SafeTx: [
        { name: "to", type: "address"},
        { name: "value", type: "uint256"},
        { name: "data", type: "bytes"},
        { name: "operation", type: "uint8"},
        { name: "safeTxGas", type: "uint256"},
        { name: "baseGas", type: "uint256"},
        { name: "gasPrice", type: "uint256"},
        { name: "gasToken", type: "address"},
        { name: "refundReceiver", type: "address"},
        { name: "nonce", type: "uint256"},
      ]
    },
    safeTx
  )
}
```